File: rtmb_java/src/com/ibm/realtime/rtmb/tests/util/ArrayOperations.java
/*******************************************************************************
 * Copyright (c) 2008 International Business Machines Corp.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    Ajith Ramanath            - initial API and implementation
 *    Radhakrishnan Thangamuthu - initial API and implementation
 *    Mike Fulton               - initial API and implementation
 *******************************************************************************/
 
package com.ibm.realtime.rtmb.tests.util;

public class ArrayOperations {
	private static final int PAGE_SIZE = 4 * 1024;

	/*
	 * Do not allow creation of instances.
	 */
	private ArrayOperations() {
	}

	/**
	 * do_array_opern() - create maximum cache misses and 
	 * force memory reads.
	 * @param m - byte array for random memory access.
	 * @param size - size of the array.
	 */

	public static int do_array_opern (byte[] m, int subIters) {
		long i;
		int size = m.length;
		int index = 0;
		assert ((size != 0) && (size > (PAGE_SIZE + 1)));
		/* Force access of another page */
		for (i = 0; i < subIters; i++) {
			index = (int) ( ((i + 1) * (100 * PAGE_SIZE) ) % m.length);
			m[index] = 'x';
		}
		return index;
	}
}

File: rtmb_cpp/src/testcase/util/ArrayOperations.cpp
/*******************************************************************************
 * Copyright (c) 2008 International Business Machines Corp.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    Ajith Ramanath            - initial API and implementation
 *    Radhakrishnan Thangamuthu - initial API and implementation
 *    Mike Fulton               - initial API and implementation
 *******************************************************************************/
 
#include <ArrayOperations.h>
#include <assert.h>

/*
 * Name    : doArrayOpern()
 * Desc    : This function is meant to create maximum
 *           cache misses and force memory reads.
 *           This test will help get a measure of memory
 *           access performance.
 * Args    : m    = array of chars.
 *           size = size of the array m.
 * Returns : 0 - on success; -1 on failure.
 */

void ArrayOperations::doArrayOpern(int size) {
	int j = 0, i = 0;
	assert ((size != 0) && (size > RTMB_PAGESIZE + 1));
	char *m = new char[size];

	/* Force access of another page */
	int j_inc = RTMB_PAGESIZE + 1;
	int end = (size / j_inc);

	for (i = 0; i < end; i++, j += j_inc) {
		if (i % 2 == 0) {
			m[j] = 'a';
		} else {
			m[end - i - 1] = 'b';
		}
	}
	delete[] m;
}
